Lib/ensurepip/__init__.py

```diff
@@ -171,7 +171,7 @@ def _bootstrap(*, root=None, upgrade=False, user=False,
         if verbosity:
             args += ["-" + "v" * verbosity]
 
-        return _run_pip([*args, "pip"], [os.fsencode(tmp_wheel_path)])
+        return _run_pip([*args, "pip"], [str(tmp_wheel_path)])
```

Comment:
@pradyunsg
Could you explain why? It's made out of a pathlib object that already returns a normal unicode string. This is then injected into a snippet of Python code executed in a subprocess.

Is there a case when it'd matter? We don't work with the OS APIs on this level so I figure it'd do nothing..

Reply:
IMO, simply because it saves people having to reason about the code. `tmp_wheel_path` is a path object, and `os.fsdecode` converts path objects to strings safely, so we know this is safe, *without* having to think too hard about edge cases like undecodable bytes, surrogate escape characters, etc.

In this case yes, both will do the same thing[^1], but why make the reader work harder than necessary?

[^1]: I assume - I haven't thought it through, which is sort of my point 🙂

Reply:
I'm questioning this exactly because I'm not sure it'd do the right or equivalent thing. I don't mind integrating the suggestion, just want to understand it first.

Reply:
`os.fsdecode` will convert a bytes value correctly, whereas `str()` won't. It's not going to matter here, but `fsdecode` is, in some sense, "more correct".

In reality, though, the way to think of it is "`fsdecode` makes a filename (in whatever form you get it) into a string". And `fsencode` is the same, but to return bytes.

Honestly, though, I'm not that fussed. For this situation, `str` seems good enough even if `fsdecode` is technically better.

Reply:
(I've had a travel day today, and am only just seeing this)

Thanks @pfmoore for the replies here -- I'm basically on the same page. `fsdecode` more clearly communicates what the intended behaviour is here and it's _technically_ better.

> I'm not sure it'd do the right or equivalent thing

It will be equivalent for str-based `Path`s. That function takes a path / path-like object and converts it to a string representation matching the filesystem encoding (which is relevant in the case of bytes paths, but we don't have that here).

It'll return a `str` unchanged.

Reply:
Makes sense. Will do.